lib/discover/global.go

```diff
@@ -0,0 +1,243 @@
+package discover
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/syncthing/protocol"
+	"github.com/syncthing/syncthing/lib/events"
+	"github.com/syncthing/syncthing/lib/sync"
+)
+
+type globalClient struct {
+	server         string
+	addrList       AddressLister
+	relayStat      RelayStatusProvider
+	announceClient *http.Client
+	queryClient    *http.Client
+	stop           chan struct{}
+	announced      bool
+	announcedMut   sync.Mutex
+}
+
+const (
+	defaultReannounceInterval  = 30 * time.Minute
+	announceErrorRetryInterval = 5 * time.Minute
+)
+
+type announcement struct {
+	Direct []string `json:"direct"`
+	Relays []Relay  `json:"relays"`
+}
+
+func NewGlobal(server string, insecure bool, cert tls.Certificate, addrList AddressLister, relayStat RelayStatusProvider) (FinderService, error) {
+	if _, err := url.Parse(server); err != nil {
+		return nil, err
+	}
+
+	// The http.Client used for announcements. It needs to have our
+	// certificate to prove our identity, and may or may not verify the server
+	// certificate depending on the insecure setting.
+	announceClient := &http.Client{
```

Comment:
Possibly. I was seeing it as a slight advantage in that we disclose less information for an eavesdropper (it's not obvious from the request exactly who is doing the asking), but then again the eavesdropper just needs to chill for an hour to get an announce instead to get the full picture.

(Also I guess the handshake is a little cheaper, but probably no big deal)